pdf-converter/client.py

```diff
@@ -0,0 +1,427 @@
+            f"{dim.depth}", f"rgb:{rep.initial}", f"png:{rep.final}"]
+
+    try:
+        proc = await asyncio.create_subprocess_exec(*cmd)
+        await wait_proc(proc)
+    except asyncio.CancelledError:
+        terminate_proc(proc)
+
+    return proc
+
+
+# TODO (?): Save->delete PNGs in loop to avoid storing all PNGs in memory.
+# TODO: Add error handling
+def combine_reps(save_path, reps):
+    with Image.open(reps[0].final) as first:
+        remaining = [Image.open(rep.final) for rep in reps[1:]]
+        first.save(save_path, "PDF", resolution=100, append_images=remaining,
+                   save_all=True)
+
+    for img in remaining:
+        img.close()
+
+
+###############################
+#           Main
+###############################
+
+async def receive(loop, proc, tmpdir):
+    procs = []
+    reps = []
+
+    try:
+        pagenums = await recv_pagenums(loop, proc)
+    except (PageError, ReceiveError):
+        raise
+
+    for page in range(1, pagenums + 1):
+        try:
+            rep, dim = await recv_rep(loop, proc, tmpdir, page)
+        except (DimensionError, ReceiveError, RepresentationError):
+            term_tasks = [terminate_proc(p) for p in procs]
+            await asyncio.gather(*term_tasks)
+            raise
+
+        reps.append(rep)
+        procs.append(await start_convert(rep, dim))
+
+    return procs, reps
+
+
+async def convert(loop, path, procs, reps):
+    for proc, rep in zip(procs, reps):
+        try:
+            await wait_proc(proc)
+        except subprocess.CalledProcessError:
+            logging.error("page conversion failed")
+            raise
+
+        await loop.run_in_executor(None, unlink, rep.initial)
+
+    save_path = path.with_suffix(".trusted.pdf")
+    await loop.run_in_executor(None, combine_reps, save_path, reps)
+
+    return save_path
+
+
+async def sanitize(loop, proc, path):
+    with TemporaryDirectory(prefix=f"qvm-sanitize-{proc.pid}-") as tmpdir:
+        try:
+            convert_procs, reps = await receive(loop, proc, tmpdir)
+        except (DimensionError, PageError, ReceiveError, RepresentationError):
+            raise
+
+        try:
+            pdf = await convert(loop, path, convert_procs, reps)
+        except (asyncio.CancelledError, subprocess.CalledProcessError):
+            for proc in convert_procs:
+                terminate_proc(proc)
+            raise
+
+    print(f"\nConverted PDF saved as: {pdf}")
+
+
+# TODO: KeyboardInterrupt
+async def run(loop, paths):
+    cmd = ["/usr/bin/qrexec-client-vm", "@dispvm", "qubes.PdfConvert"]
+    procs = []
+    send_tasks = []
+    sanitize_tasks = []
+
+    print("Sending files to Disposable VMs...")
+
+    for path in paths:
+        proc = await asyncio.create_subprocess_exec(*cmd,
+                                                    stdin=subprocess.PIPE,
+                                                    stdout=subprocess.PIPE)
```

Comment:
Ok, so you start a separate service call for each file (which is fine). But this also means, you don't strictly need to send the file size before the PDF file (the server can simply read until EOF, you can use `write_eof()` in `send_pdf()`) - which will make it compatible with the old version and greatly ease handling the upgrade.

Reply:
I literally just learned of `write_eof()` when looking into the `readuntil()` thing from above. Will implement.

> which will make it compatible with the old version and greatly ease handling the upgrade.

Woo!

Reply:
Fixed in 326e867.